### ISSUE TITLE
Inconsistent inlineing of Iterator Adaptors - Missed Optimizations

### ISSUE THREAD

**Issue by jonasbb** [rust-lang/rust#47461]
While profiling some rust code of me, I noticed that the following pattern does not optimize well:

```rust
vec![1,2,3,4]
    .into_iter()
    .map(|v| ...)
    .skip_while(|v| ...)
```

`skip_while` is implemented using `find` and `find` is implemented using `try_fold`. The functions `SkipWhile::next()` and `Iterator::find()` use the `#[inline]` annotation. The function `Map::try_fold()` does not. This means that `Map::try_fold()` will not be inlined.

I started looking at the source code and inlineing of iterators seems to follow no rule. I could not find any bug reports related to this.

* [`Filter::try_fold` is inline](https://github.com/rust-lang/rust/blob/8ff449d505728276e822ca9a80c1e7b2da8288a2/src/libcore/iter/mod.rs#L1396)
* [`Enumerate::try_fold` is inline](https://github.com/rust-lang/rust/blob/8ff449d505728276e822ca9a80c1e7b2da8288a2/src/libcore/iter/mod.rs#L1633)
* [`Rev::try_fold` is not inline](https://github.com/rust-lang/rust/blob/8ff449d505728276e822ca9a80c1e7b2da8288a2/src/libcore/iter/mod.rs#L427)

Some iterators like [`Cloned`](https://github.com/rust-lang/rust/blob/8ff449d505728276e822ca9a80c1e7b2da8288a2/src/libcore/iter/mod.rs#L515) do not have any function marked as inline. Not even `next()` is marked as inline.

The [PR introducing `try_fold`](https://github.com/rust-lang/rust/pull/45595) does not give justification why some `try_fold`s are inline and some are not.

The methods `len` and `is_empty` of `ExactSizeIterator`'s are also not marked as inlineable, even though they are always implemented as pass-through to the underlying iterator.

If desired I can prepare a pull request to mark those functions as inlineable. Is there a list of functions for the iterator traits (e.g., Iterator, ExactSizeIterator) which should be inline/not be inline?

**Comment by rkruppe** [comment]
Note that all iterator adaptors are generic and therefore instantiated in the crate using them, making them available for inlining in principle. Adding `#[inline]` in those cases just gives a hint to be slightly more aggressive about inlining than the default. Codegen units + absence of ThinLTO might break this, not sure what's the current interaction between CGUs and `#[inline]`.

We should still be consistent about how we use `#[inline]`, but absence of `#[inline]` does not generally mean that the function can't be inlined.

**Comment by varkor** [comment]
Is there an improvement in the generated instructions if you annotate `try_fold` with `#[inline]`?

**Comment by jonasbb** [comment]
@varkor I do not have a minimal example or a good testcase where I could compare the assembly. I just noticed that in profiling the `try_fold` does not occur anymore, meaning it got inlined.

@rkruppe You make a good point. I forgot that generic functions are always inlineable. I got fooled by the results.

**Comment by scottmcm** [comment]
I had no deep logic for marking or not marking try_fold methods with `#[inline]`.  Roughly I had it there when the corresponding `fold` had it, but that's it.

@jonasbb I see "anymore" in your comment.  Did the behaviour here change in a recent nightly?

**Comment by jonasbb** [comment]
@scottmcm No, it is unrelated to nightly changes. For other reasons I already have xargo setup and just played around with adding inline annotation and reprofiling.

**Comment by Pzixel** [comment]
I can confirm that I have this issue: code with `iter.copied().map(|x| something(x))` produced bad assembly while `iter.map(|x| something(*x)` was properly inlined. I've tried to create an MRE but I wasn't able to reproduce it there. I'm not an expert on compiler optimizations but I can confirm that I witnessed this behavior in the wild. 

As I said I failed to create a MRE but I can provide function source code and assembly that gets generated if needed. I suspect it has to deal something with inlining budget or some werights but I'm pretty sure I don't want to have iterators in resulting release assembly.
